### PR TITLE
Improved render bounding boxes for animated models

### PR DESCRIPTION
**Pull request by hyperlogic**
* bug fix in AABox::operator+=
* added AABox::emiggen
* added more unit tests for AABox
* Model: split collision and model mesh render items.
  Because ModelMeshRenderItems need special handling to update bounds for animated joints.
* Model: dynamically update the bound for rigidly bound animated meshes
* Rig: added access to geometryToRigTransform
* Geometry.cpp: removed unused bounds parameter in evalPartBounds
* ModelMeshPartPayload: bounds updating
   * non-animated: use existing _localBound
   * rigid bound mesh: use _localBound transformed by a single clusterMatrix joint transform
   * fully skinned mesh: use the bound that encapsulates all the _localBounds transformed by all clusterMatrix transforms.
* This also fixes an issue with eye tracking that would cause eyeballs or any mesh attached to the eyeballs to flicker.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: c671fe796694dc216a88dc71e222ff2534021444 available at: https://gist.github.com/76744fd899b015f21696<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.dmg

File: libraries/shared/src/AABox.cpp [hunk 1]
```diff
@@ -475,8 +477,15 @@ AABox AABox::clamp(float min, float max) const {
 }
 
 AABox& AABox::operator += (const glm::vec3& point) {
-    _corner = glm::min(_corner, point);
-    _scale = glm::max(_scale, point - _corner);
+
+    if (isInvalid()) {
```
Comment (samcake) [inline libraries/shared/src/AABox.cpp]: I don;t see the difference between this code  with a if /else branch and the previous one ?

File: libraries/render-utils/src/Model.h [hunk 2]
```diff
@@ -371,8 +372,12 @@ class Model : public QObject, public std::enable_shared_from_this<Model> {
     bool _renderCollisionHull;
 
 
-    QSet<std::shared_ptr<MeshPartPayload>> _renderItemsSet;
-    QMap<render::ItemID, render::PayloadPointer> _renderItems;
+    QSet<std::shared_ptr<MeshPartPayload>> _collisionRenderItemsSet;
```
Comment (samcake) [inline libraries/render-utils/src/Model.h]: Next step, getting rid of the reference to the Payloads from Model :)

File: libraries/render-utils/src/MeshPartPayload.h [hunk 3]
```diff
@@ -101,4 +101,11 @@ class ModelMeshPartPayload : public MeshPartPayload {
     bool _isBlendShaped{ false };
 };
 
+namespace render {
```
Comment (samcake) [inline libraries/render-utils/src/MeshPartPayload.h]: I don;t think you need this block of template specialization
THis is already done by the block line 67 above for the "MeshPartPayload" and since the MOdelMeshPartPayload inherit MEshPartPayload then i beleive the right thing happens there no ?

**Comment by RyanDowne** [comment]
I tested this in cellscience and HQ where I know there are animations that were not working well and they look great!

File: libraries/shared/src/AABox.cpp (hunk 1 above)
Comment (hyperlogic) [inline libraries/shared/src/AABox.cpp]: The removed code has a bug.  You'll notice it moves the corner first, before it computes the scale.  So, for example:   if you start with a box that has a corner at the origin with a scale of 1.  Then you call += with a point that is (-1, -1, -1), what would you expect the result to be?  A box with a corner at -1, and a size of 2, right?  Well the previous code would compute the corner correctly, but the scale would remain 1.   In general this was causing some of our computed bounds to have an incorrect smaller _scale.

File: libraries/render-utils/src/MeshPartPayload.h (hunk 3 above)
Comment (hyperlogic) [inline libraries/render-utils/src/MeshPartPayload.h]: I was getting link errors, when this was not present.

**Comment by hyperlogic** [comment]
Looks like I broke the debug collision geometry rendering, working on a fix.

**Comment by hyperlogic** [comment]
All better now.

**Comment by hyperlogic** [comment]
https://app.asana.com/0/45663370065758/104525008502787
https://app.asana.com/0/45663370065758/99251267924568
https://app.asana.com/0/45663370065758/103449186427166
https://app.asana.com/0/26225263936266/88419070311551

**Comment by hifi-gustavo** [comment]
Build console output for commit: f5a86666a14a93667be7f21ed64c982240c341d5 available at: https://gist.github.com/76744fd899b015f21696<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.dmg

File: libraries/model/src/model/Geometry.cpp [hunk 4]
```diff
@@ -111,7 +111,7 @@ Box Mesh::evalPartBound(int partNum) const {
     return box;
 }
 
-Box Mesh::evalPartBounds(int partStart, int partEnd, Boxes& bounds) const {
```
Comment (samcake) [inline libraries/model/src/model/Geometry.cpp]: I think my goal here was to grab the bounds for each indivual parts in the vactor and the combined bound as the return value

**Comment by cozza13** [comment]
Tested this and the objects that would be rendered out in release, now render correctly.

**Comment by samcake** [comment]
OK for me, Tony convinced me :)

File: libraries/shared/src/AABox.cpp (hunk 1 above)
Comment (samcake) [inline libraries/shared/src/AABox.cpp]: AH !

**Comment by PhilipRosedale** [comment]
I'm running with this PR and have not been able to find anything broken yet... and the windmill is now working for me too.

**Comment by bluetoe** [comment]
I tried this and it seems OK as far as basic stuff goes. I did notice a few quirks with editing stuff but checked the current production release and saw they were present with that too. The only windmill I know of here is the one at the Alpha place name and although at one time I saw it moving, today I do not.

**Comment by ZappoMan** [comment]
Note: chatting with Tony, it sounds like this PR may have a negative impact on Frame Rate... there's an independent issue that causes this new code to add more load than before.

**Comment by hifi-gustavo** [comment]
Build console output for commit: e77cf544830713c26027275b8890deb2124b28fc available at: https://gist.github.com/76744fd899b015f21696<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.dmg

**Comment by ZappoMan** [comment]
Let's hold off on merging until we know more about performance changes from this PR.

**Comment by hyperlogic** [comment]
Here are some benchmarks:

master
-----------------
[hifi://ozan/9.57438,1.04847,7.92038/0,0.427578,0,0.903979](ozan)
updatePendingChangesQueue: 2.3 ms
simrate: 51
paintGL: 17.6 ms

[hifi://demo](demo)
updatePendingChangesQueue: 0.59 ms
simrate: 75
paintGL: 10.6 ms

[hifi://playa](playa)
updatePendingChangingQueue: 0.06 ms
simrate: 75
paintGL: 6 ms

[hifi://forum](forum)
updatePendingChangingQueue: 0.09 ms
simrate: 75
paintGL: 9 ms

tony/animation-culling
-------------------

[hifi://ozan/9.57438,1.04847,7.92038/0,0.427578,0,0.903979](ozan)
updatePendingChangesQueue: 120.0 ms
simrate: 7
paintGL: 136.0 ms

[hifi://demo](demo)
updatePendingChangesQueue: 4.2 ms
simrate: 60
paintGL: 13.2 ms

[hifi://playa](playa)
updatePendingChangingQueue: 0.2 ms
simrate: 75
paintGL: 6 ms

[hifi://forum](forum)
updatePendingChangingQueue: 0.454 ms
simrate: 75
paintGL: 10.5 ms

**Comment by hyperlogic** [comment]
tl;dr: demo dropped from 75 to 60 htz in this PR.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 115fd607a0912d8c48e2a82e60271607378c3be3 available at: https://gist.github.com/76744fd899b015f21696<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.dmg

**Comment by hyperlogic** [comment]
After these changes the frame rate is back on track.  And improved for the hifi://ozan greek soldier scene.

animation-culling (pt.2)
--------------------------------
hifi://ozan/9.57438,1.04847,7.92038/0,0.427578,0,0.903979
updatePendingChangesQueue: 1.2 ms
simrate: 60
paintGL: 30.0 ms

hifi://demo
updatePendingChangesQueue: 0.8 ms
simrate: 75
paintGL: 8.3 ms

hifi://playa
updatePendingChangingQueue: 0.07 ms
simrate: 75
paintGL: 5.8 ms

hifi://forum
updatePendingChangingQueue: 0.08 ms
simrate: 75
paintGL: 9.0 ms

**Comment by hifi-gustavo** [comment]
Build console output for commit: 24ca5b3d60a7fe0919d88e74578fe71f71ede623 available at: https://gist.github.com/76744fd899b015f21696<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 80dfed77d75fa3fe2b13ce34d9e310ee1b38eebf available at: https://gist.github.com/76744fd899b015f21696<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a2c520a74a7b0ad982c95e8dcb1f0c948d8306e4 available at: https://gist.github.com/76744fd899b015f21696<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7487.dmg

**Comment by ZappoMan** [comment]
My smoke tests of this pass. I am good with this...
